pkg/collector/stats/node_stats.go

```diff
@@ -19,18 +19,175 @@ package stats
 import (
 	"fmt"
 
+	"math"
+
 	"github.com/sustainable-computing-io/kepler/pkg/config"
 	"github.com/sustainable-computing-io/kepler/pkg/node"
 	acc "github.com/sustainable-computing-io/kepler/pkg/sensors/accelerator"
 	"github.com/sustainable-computing-io/kepler/pkg/utils"
+	"k8s.io/klog/v2"
 )
 
+var (
+	// Modify Manually
+	spreadDiff           = 0.3
+	historyLength        = 10
+	energyTypeToMinSlope = map[string]float64{
+		config.IdleEnergyInCore:   0.01,
+		config.IdleEnergyInPkg:    0.3,
+		config.IdleEnergyInUnCore: 0.3,
+		config.IdleEnergyInDRAM:   0.3,
+	}
+	energyTypeToMinIntercept = map[string]float64{
+		config.IdleEnergyInCore:   0,
+		config.IdleEnergyInPkg:    0,
+		config.IdleEnergyInUnCore: 0,
+		config.IdleEnergyInDRAM:   0,
+	}
+)
+
+type IdleEnergyResult struct {
+	calculatedIdleEnergy float64
+	diff                 float64
+	history              []float64
+}
+
+type IdleEnergyCalculator struct {
+	// Min CPU TIME
+	minUtilization *Coordinate
+	// Max CPU Time
+	maxUtilization *Coordinate
+	// Min y-intercept
+	minIntercept float64
+	// Min slope
+	minSlope float64
+
+	result *IdleEnergyResult
+}
+
+func (ic *IdleEnergyCalculator) UpdateIdleEnergy(newResutilization float64, newEnergyDelta float64, maxTheoreticalCPUTime float64) {
+	klog.V(5).Infof("New Datapoint Candidate: (%f, %f)", newResutilization, newEnergyDelta)
+	klog.V(5).Infof("Current Min Utilization Datapoint: (%f, %f)", ic.minUtilization.X, ic.minUtilization.Y)
+	klog.V(5).Infof("Current Max Utilization Datapoint: (%f, %f)", ic.maxUtilization.X, ic.maxUtilization.Y)
+	klog.V(5).Infof("Current History: (%v)", ic.result.history)
+	klog.V(5).Infof("Current Spread: (%v)", ic.result.diff)
+	if newResutilization > ic.minUtilization.X && newResutilization < ic.maxUtilization.X {
+		klog.V(5).Infof("Excess Datapoint: (%f, %f)", newResutilization, newEnergyDelta)
+		// Record History
+		klog.V(5).Infof("Push Idle Energy to history")
+		appendToSliceWithSizeRestriction(&ic.result.history, historyLength, ic.result.calculatedIdleEnergy)
```

Comment:
could you please explain why this is needed? Would this be removed in the final draft ?

Reply:
This tracks consistency. One of the requirements is to make sure the idle power is consistent. So keeping track of the recent history (say the past 10 calculations) and comparing the average of the history with the current idle power can indicate if idle power is consistent (ex. percentage error difference should be less than 0.1). Note, spread and history are tracked and once the idle power passes both, the idle power is freely reported without restrictions (let me know if this should be changed)